api/envoy/config/route/v3/route_components.proto

```diff
@@ -764,10 +764,106 @@ message RouteAction {
+  //                 filters:
+  //                 - and_filter:
+  //                   filters:
+  //                   - header:
+  //                     key: content-type
+  //                     value: text/html
+  //                   - header:
+  //                     key: cookie
+  //                     value: pagetype=123
+  //                 - or_filter:
+  //                   filters:
+  //                   - header:
+  //                     key: status
+  //                     value: 301
+  //                   - header:
+  //                     key: status
+  //                     value: 302
+  //                   - header:
+  //                     key: status
+  //                     value: 307
+
+  message HeaderRewriteLiteral {
+    // String match can use prefix match and present match. In case of prefix match, the matched prefix part of
+    // the header's value is substituted. In case of present match the whole value of the header is overwritten.
+    type.matcher.v3.StringMatcher string_match = 1 [(validate.rules).message = {required: true}];
+
+    // The matched part will be replaced by the value provided in substitution.
+    string substitution = 2 [(validate.rules).string = {min_bytes: 1}];
+  }
+
+  message HeaderRewriter {
+    oneof header_rewriter_type {
+      // Describes the pattern to look for in the Header's value and what to subsitute in place of such value
+      HeaderRewriteLiteral header_rewrite_literal = 1;
+    }
+
+    // Describes which requests should be eligible for rewriting headers received from upstream. If this match
+    // criteria fails, no action would be taken on the upstream headers before sending to the end user.
+    HeaderMatcherFilter request_match_criteria = 3 [(validate.rules).message = {required: true}];
+  }
+
+  message HeaderMatcherFilter {
+    oneof filter_specifier {
+      // Only requests with a header which matches the specified
+      // HeaderMatcherFilter will pass the filter check.
+      HeaderMatcher header = 1 [(validate.rules).message = {required: true}];
+
+      // And filter
+      AndFilter and_filter = 2;
+
+      // Or filter
+      OrFilter or_filter = 3;
+    }
+  }
+
+  // Performs a logical “and” operation on the result of each filter in filters.
+  // Filters are evaluated sequentially and if one of them returns false, the
+  // filter returns false immediately.
+  message AndFilter {
+    repeated HeaderMatcherFilter filters = 1 [(validate.rules).repeated = {min_items: 2}];
+  }
+
+  // Performs a logical “or” operation on the result of each individual filter.
+  // Filters are evaluated sequentially and if one of them returns true, the
+  // filter returns true immediately.
+  message OrFilter {
+    repeated HeaderMatcherFilter filters = 2 [(validate.rules).repeated = {min_items: 2}];
+  }
+
   reserved 12, 18, 19, 16, 22, 21, 10;
 
   reserved "request_mirror_policy";
 
+  map<string, HeaderRewriter> response_headers_to_rewrite = 36;
```

Comment:
Please add docs for this field. Also, we haven't typically used a map for things like this. IMO it would be more future proof to have a list of messages, even if internally it is translated to a map? @envoyproxy/api-shepherds?